third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -368,18 +368,41 @@ pub enum FunctionAttribute {
     Persistent,
     /// During execution of the function, a module reentrancy lock is established.
     ModuleLock,
+    Pack,
+    PackVariant,
+    Unpack,
+    UnpackVariant,
+    TestVariant,
+    BorrowFieldImmutable(u8),
+    BorrowFieldMutable(u8),
 }
 
 impl FunctionAttribute {
     /// Returns true if the attributes in `with` are compatible with
     /// the attributes in `this`. Typically, `this` is an imported
     /// function handle and `with` the matching definition. Currently,
-    /// only the `Persistent` attribute is relevant for this check.
+    /// `Persistent` and all public struct related attributes are relevant for this check.
     pub fn is_compatible_with(this: &[Self], with: &[Self]) -> bool {
-        if this.contains(&FunctionAttribute::Persistent) {
-            with.contains(&FunctionAttribute::Persistent)
-        } else {
-            true
+        for attr in this {
+            if attr.cannot_be_removed() {
+                if !with.contains(attr) {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    /// Returns true if the attribute cannot be removed from the function handle once added.
+    pub fn cannot_be_removed(&self) -> bool {
+        match self {
+            FunctionAttribute::ModuleLock => false,
+            FunctionAttribute::Pack | FunctionAttribute::PackVariant => true,
```

Comment:
Why not combine all the remaining attributes (returning `true`) with `|`? (It might make it easier for compiler to deduct and optimize this to `self != ModuleLock`).